kv/kvserver: TestWedgedReplicaDetection failed

kv/kvserver.TestWedgedReplicaDetection [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/15711752?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/15711752?buildTab=artifacts#/) on master @ [5afd790501e946ef306abe2b592c5798c29c342f](https://github.com/cockroachdb/cockroach/commits/5afd790501e946ef306abe2b592c5798c29c342f):


```
=== RUN   TestWedgedReplicaDetection
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/3c632028e5b231a582105881dfd5d4b5/logTestWedgedReplicaDetection3013974570
    test_log_scope.go:81: use -show-logs to present logs inline
    client_raft_test.go:2735: expected follower to still be considered active
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/3c632028e5b231a582105881dfd5d4b5/logTestWedgedReplicaDetection3013974570
--- FAIL: TestWedgedReplicaDetection (0.91s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/replication
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWedgedReplicaDetection.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-39622

--- a/pkg/kv/kvserver/client_raft_test.go
+++ b/pkg/kv/kvserver/client_raft_test.go
@@ -2681,6 +2681,7 @@ func TestWedgedReplicaDetection(t *testing.T) {
 	require.Nil(t, err)
 	tc.WaitForValues(t, key, []int64{1, 1, 1})
 
+	// Get a handle on the leader and the follower replicas.
 	leaderRepl := tc.GetRaftLeader(t, key)
 	leaderClock := leaderRepl.Clock()
 	followerRepl := func() *kvserver.Replica {
@@ -2697,6 +2698,20 @@ func TestWedgedReplicaDetection(t *testing.T) {
 	if followerRepl == nil {
 		t.Fatal("could not get a handle on a follower replica")
 	}
+
+	// Wait for the leader replica to have three entries in its lastUpdateTimes
+	// map. It should already by this time because it was able to replicate a log
+	// entry to its two followers, but we wait here to be sure and to avoid
+	// flakiness. It is possible that the WaitForValues call above returned as
+	// soon as one of the followers appended and applied a log entry, but before
+	// its response was delivered to the leader.
+	testutils.SucceedsSoon(t, func() error {
+		lastUpdateTimes := leaderRepl.LastUpdateTimes()
+		if len(lastUpdateTimes) == 3 {
+			return nil
+		}
+		return errors.Errorf("expected leader replica to have 3 entries in lastUpdateTimes map, found %s", lastUpdateTimes)
+	})
 
 	// Lock the follower replica to prevent it from making progress from now
 	// on. NB: See TestRaftBlockedReplica/#9914 for why we use a separate

--- a/pkg/kv/kvserver/client_raft_test.go
+++ b/pkg/kv/kvserver/client_raft_test.go
@@ -2670,6 +2670,7 @@ func TestWedgedReplicaDetection(t *testing.T) {
 	// Pause the manual clock so that we can carefully control the perceived
 	// timing of the follower replica's activity.
 	manual.Pause()
+	t.Logf("paused clock at %s", manual.Now())
 
 	key := []byte("a")
 	tc.SplitRangeOrFatal(t, key)
@@ -2731,8 +2732,11 @@ func TestWedgedReplicaDetection(t *testing.T) {
 
 	// The follower should still be active.
 	followerID := followerRepl.ReplicaID()
-	if !leaderRepl.IsFollowerActiveSince(followerID, leaderClock.PhysicalTime(), inactivityThreshold) {
-		t.Fatalf("expected follower to still be considered active")
+	leaderNow := leaderClock.PhysicalTime()
+	if !leaderRepl.IsFollowerActiveSince(followerID, leaderNow, inactivityThreshold) {
+		t.Fatalf("expected follower to still be considered active; "+
+			"follower id: %d, last update times: %s, leader clock: %s",
+			followerID, leaderRepl.LastUpdateTimes(), leaderNow)
 	}
 
 	// It is possible that there are in-flight heartbeat responses from
@@ -2750,8 +2754,11 @@ func TestWedgedReplicaDetection(t *testing.T) {
 		}
 
 		// The follower should no longer be considered active.
-		if leaderRepl.IsFollowerActiveSince(followerID, leaderClock.PhysicalTime(), inactivityThreshold) {
-			return errors.New("expected follower to be considered inactive")
+		leaderNow = leaderClock.PhysicalTime()
+		if leaderRepl.IsFollowerActiveSince(followerID, leaderNow, inactivityThreshold) {
+			return errors.Errorf("expected follower to be considered inactive; "+
+				"follower id: %d, last update times: %s, leader clock: %s",
+				followerID, leaderRepl.LastUpdateTimes(), leaderNow)
 		}
 		return nil
 	})

--- a/pkg/kv/kvserver/helpers_test.go
+++ b/pkg/kv/kvserver/helpers_test.go
@@ -18,6 +18,7 @@ package kvserver
 import (
 	"context"
 	"fmt"
+	"maps"
 	"testing"
 	"time"
 
@@ -407,6 +408,12 @@ func (r *Replica) NumPendingProposals() int64 {
 	r.mu.RLock()
 	defer r.mu.RUnlock()
 	return r.numPendingProposalsRLocked()
+}
+
+func (r *Replica) LastUpdateTimes() map[roachpb.ReplicaID]time.Time {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+	return maps.Clone(r.mu.lastUpdateTimes)
 }
 
 func (r *Replica) IsFollowerActiveSince(

